### PR TITLE
Issue 575 heap pool memory metrics

### LINKED ISSUE #575: Collect Heap-pools memory metrics
Currently we only collect Heap/non-heap. Collect data for the different pools so that it can be visualized through dashboards

### PR DESCRIPTION
**Pull request by kananindzya**
closes #575 
added heap-pool memory metrics

### CODE REVIEW

File: apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java [hunk 1]
```diff
@@ -26,56 +26,91 @@
...
+        for (MemoryPoolMXBean memoryPoolMXBean : memoryPoolMXBeans) {
+            if (memoryPoolMXBean.getType() != MemoryType.HEAP) {
+                continue;
             }
-        });
-        registry.add("jvm.memory.heap.max", Labels.EMPTY, new DoubleSupplier() {
-            @Override
-            public double get() {
-                return platformMXBean.getHeapMemoryUsage().getMax();
+            if (memoryPoolMXBean.getName().contains("Eden")) {
```
[the remark this answers is not in the document]
Comment (kananindzya) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java:74]: I think it will be more informative. And this detailed information can help to add specific alerts.
Yes, I mean that they will be reported separately but with different lables(I fixed it with last commit)
image

File: apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java [hunk 2]
```diff
@@ -26,56 +26,91 @@
...
-            public double get() {
-                return platformMXBean.getHeapMemoryUsage().getUsed();
-            }
-        });
-        registry.add("jvm.memory.heap.committed", Labels.EMPTY, new DoubleSupplier() {
-            @Override
-            public double get() {
-                return platformMXBean.getHeapMemoryUsage().getCommitted();
+
+        addUsedMetric("jvm.memory.heap.used", registry, platformMXBean.getHeapMemoryUsage());
```
[the remark this answers is not in the document]
Comment (kananindzya) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java]: fixed

File: apm-agent-plugins/apm-grpc/apm-grpc-test-latest/src/test/java/co/elastic/apm/agent/grpc/latest/testapp/generated/HelloGrpc.java [hunk 3]
```diff
@@ -42,7 +42,7 @@
 /**
  */
 @javax.annotation.Generated(
-    value = "by gRPC proto compiler (version 1.29.0)",
+    value = "by gRPC proto compiler (version 1.30.0)",
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-grpc/apm-grpc-test-latest/src/test/java/co/elastic/apm/agent/grpc/latest/testapp/generated/HelloGrpc.java:45]: Why is that related?

File: apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java [hunk 4]
```diff
@@ -78,5 +81,32 @@ public double get() {
...
+                    return memoryPoolMXBean.getUsage().getCommitted();
+                }
+            });
+            registry.add("jvm.memory.heap.max", memoryPoolTags, new DoubleSupplier() {
+                @Override
+                public double get() {
+                    return memoryPoolMXBean.getUsage().getMax();
+                }
+            });
+        }
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java:110]: As these metrics nice-to-have, please surround with try/catch, and either log error on failure or debug on success. It will help us test that we don't break anything on all JVMs we use in the integration tests (including J9)

File: apm-agent-plugins/apm-grpc/apm-grpc-test-latest/src/test/java/co/elastic/apm/agent/grpc/latest/testapp/generated/HelloGrpc.java (hunk 3 above)
Comment (kananindzya) [reply apm-agent-plugins/apm-grpc/apm-grpc-test-latest/src/test/java/co/elastic/apm/agent/grpc/latest/testapp/generated/HelloGrpc.java]: It's changed after running tests, I think. Now in the master branch the value is already "1.30.0"

File: apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java (hunk 4 above)
Comment (kananindzya) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetrics.java]: added try/catch

**Comment by kananindzya** [comment]
> @kananindzya Looks great!
> A few asks:
> 
>     * Let's use `jvm.memory.heap.pool.XXX` for the metricset keys, instead of `jvm.memory.heap.XXX`
> 
>     * Please change the tag from `type` to `name`
> 
>     * Please add tests (and remove the system out)
> 
> 
> Thanks!

- changed to jvm.memory.heap.pool.XXX
- tag changed from "type" to "name"
- tests added + sout deleted :)
- added changelog

**Comment by eyalkoren** [comment]
@kananindzya we only need to change the test, so it is not platform-dependent. Would you like to do that?

**Comment by kananindzya** [comment]
> @kananindzya we only need to change the test, so it is not platform-dependent. Would you like to do that?

Hi, sorry for the delay. I unfortunately had no time last couple weeks. I am planning to do it this weekend.

**Comment by eyalkoren** [comment]
No rush, I just wanted to know this PR is still alive 🙂

**Comment by kananindzya** [comment]
> The test is dependent on JVM implementation. It should not assume specific memory pool names.

Fixed. I'm not sure if i have reached the goal :)

File: apm-agent-core/src/test/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetricsTest.java [hunk 5]
```diff
@@ -47,7 +53,23 @@ void testMetrics() {
...
+            final Labels spaceLabel = Labels.Mutable.of("name", memoryPoolName);
+            assertThat(registry.getGaugeValue("jvm.memory.heap.pool.used", spaceLabel)).isNotZero();
+            assertThat(registry.getGaugeValue("jvm.memory.heap.pool.used", spaceLabel)).isNotNaN();
+            assertThat(registry.getGaugeValue("jvm.memory.heap.pool.committed", spaceLabel)).isNotZero();
+            assertThat(registry.getGaugeValue("jvm.memory.heap.pool.committed", spaceLabel)).isNotNaN();
+            assertThat(registry.getGaugeValue("jvm.memory.heap.pool.max", spaceLabel)).isNotZero();
+            assertThat(registry.getGaugeValue("jvm.memory.heap.pool.max", spaceLabel)).isNotNaN();
+        }
         final long[] longs = new long[1000000];
         System.out.println(registry.toString());
```
[the remark this answers is not in the document]
Comment (kananindzya) [reply apm-agent-core/src/test/java/co/elastic/apm/agent/metrics/builtin/JvmMemoryMetricsTest.java]: removed
